package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,943 @@
+                                    elif dd == 1:
+                                        shift = j
+                                    else:
+                                        shift = k
+
+                                    if shift:
+                                        d2new_c = 0
+
+                                        for d in range(DIM):
+                                            d2new_c += (pos[d] + trial[d] - shift*box[dd, d])**2
+
+                                        if d2new_c <= BOX_MARGIN*d2new:
+                                            use = False
+
+                                if use: # Accept this shift vector.
+                                    if self.c_pbcbox.ntric_vec >= MAX_NTRICVEC:
+                                        with gil:
+                                            print("\nWARNING: Found more than %d triclinic "
+                                                  "correction vectors, ignoring some."
+                                                  % MAX_NTRICVEC)
+                                            print("  There is probably something wrong with "
+                                                  "your box.")
+                                            print(np.array(box))
+
+                                            for i in range(self.c_pbcbox.ntric_vec):
+                                                print(" -> shift #{}: [{}, {}, {}]".format(i+1,
+                                                                                           self.c_pbcbox.tric_shift[i][XX],
+                                                                                           self.c_pbcbox.tric_shift[i][YY],
+                                                                                           self.c_pbcbox.tric_shift[i][ZZ]))
+                                    else:
+                                        for d in range(DIM):
+                                            self.c_pbcbox.tric_vec[self.c_pbcbox.ntric_vec][d] = \
+                                                trial[d]
+                                        self.c_pbcbox.tric_shift[self.c_pbcbox.ntric_vec][XX] = i
+                                        self.c_pbcbox.tric_shift[self.c_pbcbox.ntric_vec][YY] = j
+                                        self.c_pbcbox.tric_shift[self.c_pbcbox.ntric_vec][ZZ] = k
+                                        self.c_pbcbox.ntric_vec += 1
+
+
+    def update(self, real[:,::1] box):
+        if box.shape[0] != DIM or box.shape[1] != DIM:
+            raise ValueError("Box must be a %i x %i matrix. (shape: %i x %i)" %
+                             (DIM, DIM, box.shape[0], box.shape[1]))
+        if (box[XX, XX] == 0) or (box[YY, YY] == 0) or (box[ZZ, ZZ] == 0):
+            raise ValueError("Box does not correspond to PBC=xyz")
+        self.fast_update(box)
+
+
+    cdef void fast_pbc_dx(self, rvec ref, rvec other, rvec dx) nogil:

```

Comment:
So this is a different algorithm to what we use for all other minimum image calculations.  It looks like it doesn't have to search all images and pick the minimum (which is what our triclinic pbc has to do) in which case if this works just as well then it's an improvement.

Either way, I don't like having a different minimum image solution in this class, we should use the same throughout MDA.  So either this is better, in which case implement it everywhere, or it's wrong, in which case use the one in `calc_distances.h`

Reply:
It would be good to test this against the calc_distances implementation for triclinic cells with small angles and random data. I do understand how this works and that it can be faster then the full image search It would also work independent of box type. As @richardjgowers says this should use the code in `calc_distances.h` and replace it if this turns out to be faster.